src/Servers/Kestrel/Core/src/Internal/Http/HttpRequestStream.cs

```diff
@@ -140,11 +142,77 @@ private ValueTask<int> ReadAsyncWrapper(Memory<byte> destination, CancellationTo
             }
         }
 
-        private async ValueTask<int> ReadAsyncInternal(Memory<byte> buffer, CancellationToken cancellationToken)
+        private ValueTask<int> ReadAsyncInternal(Memory<byte> buffer, CancellationToken cancellationToken)
+        {
+            if (_asyncReader?.InProgress ?? false)
+            {
+                return new ValueTask<int>(Task.FromException<int>(new InvalidOperationException("Concurrent reads are not supported")));
+            }
+
+            try
+            {
+                while (true)
+                {
+                    if (!_pipeReader.TryRead(out var result))
+                    {
+                        break;
+                    }
+
+                    if (result.IsCanceled)
+                    {
+                        throw new OperationCanceledException("The read was canceled");
+                    }
+
+                    var readableBuffer = result.Buffer;
+                    var readableBufferLength = readableBuffer.Length;
+
+                    var consumed = readableBuffer.End;
+                    var actual = 0;
+                    try
+                    {
+                        if (readableBufferLength != 0)
+                        {
+                            actual = (int)Math.Min(readableBufferLength, buffer.Length);
+
+                            var slice = actual == readableBufferLength ? readableBuffer : readableBuffer.Slice(0, actual);
+                            consumed = slice.End;
+                            slice.CopyTo(buffer.Span);
+
+                            return new ValueTask<int>(actual);
+                        }
+
+                        if (result.IsCompleted)
+                        {
+                            return new ValueTask<int>(0);
+                        }
+                    }
+                    finally
+                    {
+                        _pipeReader.AdvanceTo(consumed);
+                    }
+                }
+            }
+            catch (Exception ex)
+            {
+                return new ValueTask<int>(Task.FromException<int>(ex));
+            }
+
+            var asyncReader = _asyncReader;
+            if (asyncReader is null)
+            {
+                _asyncReader = asyncReader = new AsyncEnumerableReader();
```

Comment:
For my own knowledge, storing a class variable as a local here is worth it even though you have to assign both _asyncReader and asyncReader?

Reply:
It has a high ratio of reads to the write; so using it in a register is better than looking it up from L1 cache each time

Reply:
The Jit may decide to hoist the field to a register; but is less likely to than a local